src/lib/arch/riscv/Instruction_execute.cc

```diff
@@ -827,6 +841,642 @@ void Instruction::execute() {
+        // them to be rounded incorrectly. Rounding mode needs to be set on
+        // commit and all subsequent instructions must be performed with updated
+        // mode. e.g. this should be performed atomically
+
+        switch (operands[0].get<uint64_t>()) {
+          case 0:
+            fesetround(FE_TONEAREST);
+            break;
+          case 1:
+            fesetround(FE_TOWARDZERO);
+            break;
+          case 2:
+            fesetround(FE_DOWNWARD);
+            break;
+          case 3:
+            fesetround(FE_UPWARD);
+            break;
+          case 4:
+            fesetround(FE_TONEAREST);
+            break;
+          default:
+            // Invalid Case
+            // TODO any subsequent attempt to execute a floating-point operation
+            // with a dynamic rounding mode will raise an illegal instruction
+            // exception.
+            std::cerr << "[SimEng:RISC-V:Execute] Invalid rounding mode"
+                      << std::endl;
+            break;
+        }
+      } else {
+        //        std::cerr << "[SimEng:RISC-V:Execute] Unknown system register:
+        //        "
+        //                  << metadata.operands[1].reg << std::endl;
+      }
+      results[0] = RegisterValue(0, 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRWI: {  // CSRRWI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRS: {  // CSRRS rd,csr,rs1
+      // dummy implementation needs capstone update to recognise system
+      // registers
+
+      results[0] = RegisterValue(static_cast<uint64_t>(0), 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRSI: {  // CSRRSI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRC: {  // CSRRC rd,csr,rs1
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRCI: {  // CSRRCI rd,csr,imm
+      executionNYI();
+      break;
+    }
+
+      // TODO "Apart from transfer operations ... all other floating-point
+      // operations on narrower n-bit operations, n < FLEN, check if the input
+      // operands are correctly NaN-boxed, i.e., all upper FLEN−n bits are 1. If
+      // so, the n least-significant bits of the input are used as the input
+      // value, otherwise the input value is treated as an n-bit canonical NaN."
+
+      // TODO need to take rounding mode into account as these can be set per
+      // instruction https://en.cppreference.com/w/cpp/numeric/fenv/FE_round
+      //      std::fesetround(FE_TONEAREST);
+
+      // Single-Precision Floating-Point (F)
+      // Double-Precision Floating-Point (D)
+    case Opcode::RISCV_FSD: {  // FSD rs1,rs2,imm
+      memoryData[0] = operands[0];
+      break;
+    }
+    case Opcode::RISCV_FSW: {  // FSW rs1,rs2,imm
+      memoryData[0] = operands[0];
+      break;
+    }
+    case Opcode::RISCV_FLD: {  // FLD rd,rs1,imm
+      // Note: elements of memory data are RegisterValue's
```

Comment:
Comment not needed as `memoryData` is not templated to be anything other than a vector of RegisterValues